examples/tls-server.rs

```diff
@@ -0,0 +1,311 @@
+            holding_registers: Arc::new(Mutex::new(holding_registers)),
+        }
+    }
+}
+
+/// Helper function implementing reading registers from a HashMap.
+fn register_read(
+    registers: &HashMap<u16, u16>,
+    addr: u16,
+    cnt: u16,
+) -> Result<Vec<u16>, std::io::Error> {
+    let mut response_values = vec![0; cnt.into()];
+    for i in 0..cnt {
+        let reg_addr = addr + i;
+        if let Some(r) = registers.get(&reg_addr) {
+            response_values[i as usize] = *r;
+        } else {
+            // TODO: Return a Modbus Exception response `IllegalDataAddress` https://github.com/slowtec/tokio-modbus/issues/165
+            println!("SERVER: Exception::IllegalDataAddress");
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::AddrNotAvailable,
+                format!("no register at address {reg_addr}"),
+            ));
+        }
+    }
+
+    Ok(response_values)
+}
+
+/// Write a holding register. Used by both the write single register
+/// and write multiple registers requests.
+fn register_write(
+    registers: &mut HashMap<u16, u16>,
+    addr: u16,
+    values: &[u16],
+) -> Result<(), std::io::Error> {
+    for (i, value) in values.iter().enumerate() {
+        let reg_addr = addr + i as u16;
+        if let Some(r) = registers.get_mut(&reg_addr) {
+            *r = *value;
+        } else {
+            // TODO: Return a Modbus Exception response `IllegalDataAddress` https://github.com/slowtec/tokio-modbus/issues/165
+            println!("SERVER: Exception::IllegalDataAddress");
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::AddrNotAvailable,
+                format!("no register at address {reg_addr}"),
+            ));
+        }
+    }
+
+    Ok(())
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let socket_addr = "127.0.0.1:8802".parse().unwrap();
+
+    tokio::select! {
+        _ = server_context(socket_addr) => unreachable!(),
+        _ = client_context(socket_addr) => println!("Exiting"),
+    }
+
+    Ok(())
+}
+
+
+async fn server_context(socket_addr: SocketAddr) -> anyhow::Result<()> {
+    println!("Starting up server on {socket_addr}");
+    let listener = TcpListener::bind(socket_addr).await?;
+    let server = Server::new(listener);
+
+    let on_connected = |stream, _socket_addr| async move {
+        let cert_path = Path::new("./pki/server.pem");
+        let key_path = Path::new("./pki/server.key");
+        let certs = load_certs(cert_path).unwrap();
+        let mut keys = load_keys(key_path, None).unwrap();
+        let config = rustls::ServerConfig::builder()
+            .with_safe_defaults()
+            .with_no_client_auth()
+            .with_single_cert(certs, keys.remove(0))
+            .map_err(|err| io::Error::new(io::ErrorKind::InvalidInput, err)).unwrap();
+        let acceptor = TlsAcceptor::from(Arc::new(config));
+
+        let service = Some(ExampleService::new());
```

Comment:
Enclosing in `Some` here and later using `map()` is pointless and confusing. Please simplify the code.

Reply:
Done